README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# validinsee

<!-- badges: start -->

![GitHub top
language](https://img.shields.io/github/languages/top/InseeFrLab/validinsee)
[![R-CMD-check](https://github.com/InseeFrLab/validinsee/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/InseeFrLab/validinsee/actions/workflows/R-CMD-check.yaml)
[![codecov](https://codecov.io/gh/InseeFrLab/validinsee/branch/main/graph/badge.svg)](https://codecov.io/gh/InseeFrLab/validinsee)
[![CRAN
status](https://www.r-pkg.org/badges/version/validinsee)](https://cran.r-project.org/package=validinsee)
<!-- badges: end -->

Le package R **validinsee** permet de valider certaines données
produites par l’Insee, en particulier la conformité de numéros Siren ou
Siret.

## Installation

``` r
# install.packages("remotes")
remotes::install_github("InseeFrLab/validinsee")
```

## Chargement

``` r
library(validinsee)
```

## Validité de numéros SIREN ou SIRET

La fonction `validation_siren` vérifie si les éléments d’un vecteur sont
des **SIREN** valides (identifiant de l’entreprise formé de 9
caractères) :

``` r
validation_siren(c("200034528", "200034582", NA))
#> [1]  TRUE FALSE    NA
```

La fonction `validation_siret` vérifie si les éléments d’un vecteur sont
des **SIRET** valides (identifiant de l’établissement formé de 14
caractères) :

``` r
validation_siret(c("20003452800014", "20003452800041", NA))
#> [1]  TRUE FALSE    NA
```

**Règles de validation**

La vérification se base sur la [formule de
Luhn](https://fr.wikipedia.org/wiki/Formule_de_Luhn).

-   un SIREN est valide si sa somme de Luhn est un multipe de 10 ;

-   un SIRET est valide si sa somme de Luhn est un multipe de 10 et si
    la somme de Luhn de son SIREN (9 premiers chiffres) est un multiple
    de 10 ;

-   un SIRET de *La Poste* (SIREN 356000000) ne respectant pas la règle
    ci-dessus fait l’objet d’un traitement différencié : il est valide
    si la somme des chiffres le composant est un multiple de 5.


NEWS.md
## validinsee 0.2.0

[30/10/2023]

* ajout d'alias `validation_siren` et `validation_siret`
* amélioration documentation

## validinsee 0.1.0

[24/10/2023]

* fonction `validation_sirene`


tests/testthat.R
library(testthat)
library(validinsee)

test_check("validinsee")


tests/testthat/test-validation_sirene.R
test_that("somme_luhn", {

  expect_equal(
    somme_luhn(matrix(1:2, ncol = 2)),
    4
  )

  expect_equal(
    somme_luhn(matrix(c(9L, 7L), ncol = 2)),
    16
  )

  expect_equal(
    somme_luhn(matrix(c(1L, 2L, 9L, 7L), ncol = 2, byrow = TRUE)),
    c(4, 16)
  )

  expect_equal(
    somme_luhn(matrix(c(9L, 7L, 3L), ncol = 3, byrow = TRUE)),
    17
  )

  expect_equal(
    somme_luhn(
      matrix(
        c(9L, 7L, 2L, 4L, 8L, 7L, 0L, 8L, 6L,
          9L, 2L, 7L, 4L, 8L, 7L, 0L, 8L, 6L), # inversion pos 2 et 3
        ncol = 9, byrow = TRUE
      )
    ),
    c(50, 54)
  )

})

test_that("validation_sirene (type siren)", {

  expect_warning(
    validation_sirene(c("1", NA), type = "siren"),
    "(essayer `type = \"siret\"` ?)"
  )

  expect_equal(
    validation_sirene(c("100000000", NA), type = "siren"),
    c(FALSE, NA)
  )

  expect_equal(
    validation_sirene(c("100000000", "43", "a"), type = "siren"),
    c(FALSE, FALSE, FALSE)
  )

  expect_equal(
    validation_sirene(c(NA, NA), type = "siren"),
    c(NA, NA)
  )

  expect_equal(
    validation_sirene(c("100000009", "100000000", NA), type = "siren"),
    c(TRUE, FALSE, NA)
  )

  expect_equal(
    validation_sirene(c("000000083", "000000080", NA), type = "siren"),
    c(TRUE, FALSE, NA)
  )

})

test_that("validation_sirene (type siret)", {

  expect_warning(
    validation_sirene(c("1", NA), type = "siret"),
    "(essayer `type = \"siren\"` ?)"
  )

  expect_equal(
    validation_sirene(c("77984790400034", "42", NA), type = "siret"),
    c(FALSE, FALSE, NA)
  )

  expect_equal(
    validation_sirene(c("77984790400033", "77984790400034", NA), type = "siret"),
    c(TRUE, FALSE, NA)
  )

  # La Poste
  expect_equal(
    validation_sirene("35600000000048", type = "siret"), # règle générale
    TRUE
  )
  expect_equal(
    validation_sirene("35600000027250", type = "siret"), # règle spécifique La Poste
    TRUE
  )
  expect_equal(
    validation_sirene("35600000000049", type = "siret"), # aucune des deux règles
    FALSE
  )

})

test_that("alias", {

  expect_equal(
    validation_siren(c("100000009", "100000000", "42", NA)),
    c(TRUE, FALSE, FALSE, NA)
  )

  expect_equal(
    validation_siret(c("77984790400033", "77984790400034", "42", NA)),
    c(TRUE, FALSE, FALSE, NA)
  )

})


R/validation_sirene.R
# https://fr.wikipedia.org/wiki/Formule_de_Luhn
# @param id_int matrice des id (chaque cellule est un chiffre unique)

somme_luhn <- function(id_int) {

  stopifnot(
    is.matrix(id_int) &&
    is.integer(id_int) &&
    all(id_int >= 0) &&
    all(id_int <= 9)
  )

  # inverse l'ordre des colonnes
  m <- id_int[ , ncol(id_int):1, drop = FALSE]

  # multiplie les éléments des colonnes paires par 2
  col_paires <- which(seq(ncol(m)) %% 2 == 0)
  m[ , col_paires] <- m[ , col_paires] * 2
  # fait la somme des chiffres si supérieur à 9
  m[m > 9] <- m[m > 9] - 9

  rowSums(m)

}

#' Validité de SIRET ou SIREN
#'
#' Vérifie la validité d'un vecteur de SIRET ou de SIREN.
#'
#' La vérification se base sur la
#' [formule de Luhn](https://fr.wikipedia.org/wiki/Formule_de_Luhn).
#'
#' - un SIREN est valide si sa somme de Luhn est un multipe de 10 ;
#'
#' - un SIRET est valide si sa somme de Luhn est un multipe de 10 et si la somme
#'   de Luhn de son SIREN (9 premiers chiffres) est un multiple de 10 ;
#'
#' - un SIRET de _La Poste_ (SIREN 356000000) ne respectant pas la règle
#'   ci-dessus fait l'objet d'un traitement différencié : il est valide si la
#'   somme des chiffres le composant est un multiple de 5.
#'
#' Les fonctions `validation_siren` et `validation_siret` permettent d'appeler
#' `validation_sirene` sans avoir à préciser le type de numéro.
#'
#' @param id un vecteur caractère de SIRET, ou de SIREN si `type = "siren"`.
#' @param type par défaut `"siret"` (chaîne de caractères de 14 chiffres).
#'   Autre valeur possible `"siren"` (9 chiffres).
#' @param warn `FALSE` pour désactiver d'éventuels warnings.
#'
#' @return Un vecteur booléen de même longueur que `id`. Une valeur manquante
#'   produira une valeur manquante en retour.
#' @export
#'
#' @examples
#' validation_sirene(c("20003452800014", "20003452800041", "a", NA)) # SIRET (14 chiffres)
#' validation_sirene(c("200034528", "200034582", "a", NA), type = "siren")
#'
#' # utiliser les alias pour ne pas renseigner `type`
#' validation_siret(c("20003452800014", "20003452800041", "a", NA))
#' validation_siren(c("200034528", "200034582", "a", NA))

validation_sirene <- function(id, type = c("siret", "siren"), warn = TRUE) {

  if (all(is.na(id))) return(as.logical(id))

  type <- match.arg(type)

  res <- rep(NA, length(id))

  if (type == "siret")
    looks_ok <- grepl("^\\d{14}$", id)
  else
    looks_ok <- grepl("^\\d{9}$", id)

  if (warn && !any(looks_ok))
    warning(
      "aucun id ne comporte le bon nombre de chiffres ",
      sprintf("(essayer `type = \"%s\"` ?)", chartr("nt", "tn", type))
    )

  res[!looks_ok & !is.na(id)] <- FALSE
  if (all(!looks_ok)) return(res)

  id <- id[looks_ok]

  # transformation en matrice numérique pour somme_luhn()
  mat_id_num <-
    matrix(
      as.integer(unlist(strsplit(id, ""))),
      nrow = length(id),
      byrow = TRUE
    )

  luhn_ok <- somme_luhn(mat_id_num) %% 10 == 0

  if (type == "siret") {

    # le SIREN a sa propre clé de contrôle

    sommes_luhn_siren <- somme_luhn(mat_id_num[ , 1:9, drop = FALSE])
    luhn_ok_siren <- sommes_luhn_siren %% 10 == 0
    luhn_ok <- luhn_ok & luhn_ok_siren

    # Critère différent pour établissements La Poste (somme simple multiple de 5)
    # https://groupes.renater.fr/sympa/arc/pstage-utilisateurs/2014-02/msg00035.html

    poste_invalides <- substr(id, 1, 9) == "356000000" & !luhn_ok
    if (any(poste_invalides)) {
      luhn_ok[poste_invalides] <- rowSums(mat_id_num[poste_invalides, , drop = FALSE]) %% 5 == 0
    }

  }

  res[looks_ok] <- luhn_ok

  res

}

#' @rdname validation_sirene
#' @export

validation_siren <- function(id)
  validation_sirene(id, type = "siren", warn = FALSE)

#' @rdname validation_sirene
#' @export

validation_siret <- function(id)
  validation_sirene(id, type = "siret", warn = FALSE)
